aries_vcx/src/handlers/issuance/credential_def/mod.rs

```diff
@@ -186,43 +147,99 @@ impl CredentialDef {
+            }
+            _ => return {
+                Err(VcxError::from_msg(VcxErrorKind::RevRegDefNotFound,
+                                       "No revocation registry definitions associated with this credential definition"
+                ))
+            }
+        };
+        self.rev_reg = Some(new_rev_reg.clone());
+        trace!("rotate_rev_reg_def <<< new_rev_reg_def: {:?}", new_rev_reg);
+        Ok(new_rev_reg)
+    }
 
-                        anoncreds::publish_rev_reg_def(&self.issuer_did, &rev_reg.rev_reg_def)
-                            .map_err(|err| err.map(VcxErrorKind::CreateCredDef, "Cannot create CredentialDefinition"))?;
+    pub fn publish_revocation_primitives(&mut self, tails_url: &str) -> VcxResult<()> {
+        warn!("publish_revocation_primitives >>> tails_url: {}", tails_url);
+        self.publish_built_rev_reg_def(tails_url)?;
+        self.publish_built_rev_reg_delta()?;
+        Ok(())
+    }
 
-                        anoncreds::publish_rev_reg_delta(&self.issuer_did, &rev_reg.rev_reg_id, &rev_reg.rev_reg_entry)
-                            .map_err(|err| err.map(VcxErrorKind::InvalidRevocationEntry, "Cannot post RevocationEntry"))?;
+    pub fn has_pending_revocations_primitives_to_be_published(&self) -> bool {
+        match &self.rev_reg {
+            None => false,
+            Some(rev_reg) => {
+                rev_reg.state_rev_reg_def == PublicEntityStateType::Built
+                    ||  rev_reg.state_rev_reg_delta == PublicEntityStateType::Built
+            }
+        }
+    }
 
-                    }
-                    _ => ()
+    pub fn publish_built_rev_reg_def(&mut self, tails_url: &str) -> VcxResult<()> {
+        match &self.rev_reg {
+            Some(rev_reg) => {
+                if rev_reg.state_rev_reg_def == PublicEntityStateType::Built {
+                    let new_rev_reg = rev_reg.clone().publish_rev_reg_def(&self.issuer_did, tails_url)?;
+                    self.rev_reg = Some(new_rev_reg);
+                } else {
+                    info!("No unpublished revocation registry definition found. Nothing to publish.")
                 }
             }
-            Err(err) => return Err(err)
-        };
+            _ => return {
+                Err(VcxError::from_msg(VcxErrorKind::NotReady,
+                                       "Tried to publish revocation primitives, but this credential definition is not revocable."
+                ))
+            }
+        }
+        Ok(())
+    }
 
-        Ok(
-            Self {
-                state: PublicEntityStateType::Published,
-                ..self
+    pub fn publish_built_rev_reg_delta(&mut self) -> VcxResult<()> {
+        match &self.rev_reg {
+            Some(rev_reg) => {
+                if rev_reg.state_rev_reg_delta == PublicEntityStateType::Built {
+                    let new_rev_reg = rev_reg.clone().publish_rev_reg_delta(&self.issuer_did)?;
+                    self.rev_reg = Some(new_rev_reg);
```

Comment:
Can be written on a single line as `self.rev_reg = Some(rev_reg.clone().publish_rev_reg_delta(&self.issuer_did)?);`.

Reply:
applied